---
title: "differential analysis"
author: "Andrea"
date: "24/09/06"
output: html_document
---

# adapted from Dane's P300_BE_Screen_edgeR.Rmd script


#--------------------------------
# SETUP
#--------------------------------

## library imports
```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(edgeR)
library(bartools)
library(clusterProfiler)
library(RColorBrewer)
library(drawProteins)
```

# set dirs
```{r}
project.dir <- "/dawson_genomics/Projects/BRD4/baseEditor_240808"
counts.dir <- file.path(project.dir, "guide_counts")
plots.dir <- file.path(project.dir, "plots")
edger.dir <- file.path(project.dir, "edgeR")
ref.dir <- file.path(project.dir, "reference_files")
```

## import counts files
```{r}
raw.counts <- read.delim(
  file.path(counts.dir, "all_guideCounts.txt"), 
  header = TRUE, 
  sep = "\t", 
  row.names = 1
  )
```

## generate metadata
```{r}
samp.info <- str_split(colnames(raw.counts)[-1], pattern = "\\.", simplify = T)
metadata <- data.frame(
  "BE" = samp.info[, 1], 
  "Treatment" = ifelse(samp.info[, 2] == "D", "DMSO", ifelse(samp.info[, 2] == "9", "965A", "None")),
  "Days" = ifelse(samp.info[, 3] == "7", "D7", ifelse(samp.info[, 3] == "14", "D14", "T0")),
  "Replicate" = samp.info[, 4],
  row.names = colnames(raw.counts)[-1]
  )
metadata$Group <- paste(metadata$BE, metadata$Treatment, metadata$Days, sep = "_")
```

## setup sgRNA info table
```{r}
guides <- read_csv("/dawson_genomics/Projects/CCS1477/baseEditor_240529/reference_files/Guides_BRD4_VHL.csv")
g.info <- str_split(guides$`Guide ID`, pattern = "_", simplify = TRUE)
guide.ann <- data.frame(
  "Gene" = g.info[, 1],
  "ID" = guides$`Guide ID`,
  "Sequence" = guides$`Guide sequence`,
  "Rank" = g.info[, 4]
)
rownames(guide.ann) <- guide.ann$ID

guide.ann <- guide.ann[rownames(raw.counts), ]
dim(guide.ann)

table(rownames(raw.counts) == rownames(guide.ann))
```

## incorporate info from design tool (CRISPick)
```{r}
# cp.all crispick data from previous script saved in env
dim(cp.all)
cp.all$aa.pos <- round(cp.all$Target.Cut.Length/3, digits = 0)
target.aa.pos <- cp.all %>% dplyr::select(sgRNA.Sequence, Target.Cut.Length, aa.pos)
```


## setup DGEList object
```{r}
dge <- DGEList(
  counts = as.matrix(raw.counts[, -1]), 
  samples = metadata, 
  group = metadata$Group,
  genes = guide.ann, 
  remove.zeros = TRUE
  )
```

## Filter sgRNAs
```{r}
table(dge$samples$group)
# Filter sgRNAs with low counts Need a CPM greater than 5 in 6 or more samples
keep = rowSums(dge$counts > 5) > 5
dge = dge[keep, ]
dim(dge)
```

#--------------------------------
# QC & FILTERING
#--------------------------------

## reads per sample
```{r}
# Set up group colours
cols <- dge$samples$group
levels(cols) <- brewer.pal(nlevels(cols), "Paired") 
cols <- as.character(cols)
# Plot number of sgRNAs that could be matched per sample 
# and total for each sgRNA across all samples
pdf(file.path(plots.dir, "barplots_screen_counts_per_index.pdf"), width = 12, height = 8)
barplot(
  colSums(dge$counts), 
  las = 2, 
  main = "Counts per index", 
  col = cols, 
  cex.names = 0.5,
  cex.axis = 0.8
  )
barplot(
  rowSums(dge$counts), 
  las = 2, 
  main = "Counts per sgRNA",
  cex.names = 0.1, 
  cex.axis = 0.8
  )
dev.off()
```

## MDS plots - all samples
```{r}
# Make an MDS plot to visualise relationships between replicates 
pdf(file.path(plots.dir, "MDS_by_group.pdf"))
plotMDS(dge, col = cols, main = "MDS Plot")
dev.off()
```

## save log2 CPM counts
```{r}
dge.cpm <- edgeR::cpm(dge, log = TRUE)
write.table(
  dge.cpm, 
  file.path(edger.dir, "logCPM_counts_all.txt"), 
  sep = "\t",
  col.names = NA,
  row.names = TRUE,
  quote = FALSE
  )
```

## bartools QC
```{r}
plotBarcodeCorrelation(dge, method = "spearman")
plotBarcodeCorrelation(dge, method = "pearson")
```

#--------------------------------
# ANALYSIS
#--------------------------------

## setup DGE model
```{r}
# Begin differential representation analysis We will use GLMs in edgeR in this 
# case since there are more than 2 groups Set up design matrix for GLM
design = model.matrix(~0 + Group, data = dge$samples)
colnames(design) <- gsub("Group", colnames(design), replacement = "")
design
```

## Estimate dispersions
```{r}
xglm = estimateDisp(dge, design, tagwise = TRUE)
sqrt(xglm$common.disp)
plotBCV(xglm, main = "BCV Plot")
fit = glmFit(xglm, design)

# save BCV plot
pdf(file.path(plots.dir, "BCVplot_all.pdf"))
plotBCV(xglm, main = "BCV Plot")
dev.off()
```

## setup contrasts 
```{r}
contr.mat <- makeContrasts(
  ABE_D7_965AvDMSO = ABE_965A_D7-ABE_DMSO_D7,
  ABE_D14_965AvDMSO = ABE_965A_D14-ABE_DMSO_D14,
  ABE_DMSO_D14vT0 = ABE_DMSO_D14-ABE_None_T0,
  ABE_DMSO_D7vT0 = ABE_DMSO_D7-ABE_None_T0,
  BE4_D7_965AvDMSO = BE4_965A_D7-BE4_DMSO_D7,
  BE4_D14_965AvDMSO = BE4_965A_D14-BE4_DMSO_D7,
  BE4_DMSO_D14vT0 = BE4_DMSO_D14-BE4_None_T0,
  BE4_DMSO_D7vT0 = BE4_DMSO_D7-BE4_None_T0,
  levels = design
  )
contr.mat
```

## LRT and smear plot for all 
```{r}
# inititate lists to store tables for each comp
lrt.list <- list()
top4.list <- list()

for(con.i in 1:length(colnames(contr.mat))){
  # Carry out Likelihood ratio test
  lrt.list[[con.i]] = glmLRT(fit, contrast = contr.mat[, con.i])
  # Show top ranked sgRNAs
  topTags(lrt.list[[con.i]], n = 50, sort.by = "logFC")
  # Select BRD4 sgRNAs with FDR < 0.0001 to highlight on plot
  top4.list[[con.i]] = topTags(lrt.list[[con.i]], n = Inf)
  top4ids = rownames(top4.list[[con.i]]$table[abs(top4.list[[con.i]]$table$logFC) >= 1 & top4.list[[con.i]]$table$Gene == "BRD4", ])
  
  # Plot logFC versus logCPM
  pdf(file.path(plots.dir, paste0("smearPlot_", colnames(contr.mat)[con.i], ".pdf")))
  plotSmear(
    lrt.list[[con.i]], 
    de.tags = top4ids,
    pch = 20, 
    cex = 0.6, 
    ylim = c(-7, 7),
    main = paste(colnames(contr.mat)[con.i], "BRD4 highlighted", sep = "\n")
    )
  abline(h = c(-1, 0, 1), col = c("dodgerblue", "yellow", "dodgerblue"), lty = 2)
  dev.off()
  
  # sort top table and write out
  top4.list[[con.i]]$table %>% arrange(FDR)
  write.table(
    top4.list[[con.i]]$table,
    file.path(edger.dir, paste0(colnames(contr.mat)[con.i], "_LRT_topTable.txt")),
    sep = "\t",
    quote = FALSE,
    col.names = TRUE,
    row.names = FALSE
  )
}

# add names for referencing
names(lrt.list) <- colnames(contr.mat)
```

## incorporate aa. pos and save logFC and Z.score ordered output
```{r}
# inititate list
top.brd4.list <- list()

for(con.i in 1:length(colnames(contr.mat))){
  # calc Z score for BRD4 guides
  top.brd4.list[[con.i]] <- as.data.frame(top4.list[[con.i]]) %>% dplyr::filter(Gene == "BRD4")
  top.brd4.list[[con.i]]$Z.score <- (top.brd4.list[[con.i]]$logFC - mean(top.brd4.list[[con.i]]$logFC))/sd(top.brd4.list[[con.i]]$logFC)
  dim(top.brd4.list[[con.i]]) 
  # summary stats on Z score distribution
  summary(top.brd4.list[[con.i]]$Z.score)
  plot(density(top.brd4.list[[con.i]]$Z.score), main = colnames(contr.mat)[con.i])
  plot(density(top.brd4.list[[con.i]]$logFC), main = colnames(contr.mat)[con.i])
  # include opposite sign for protiler
  top.brd4.list[[con.i]]$neg.Z.score <- top.brd4.list[[con.i]]$Z.score * -1
  plot(density(top.brd4.list[[con.i]]$neg.Z.score), main = colnames(contr.mat)[con.i])
  # join aa pos info into BRD4 table
  colnames(target.aa.pos)[1] <- "Sequence"
  dim(target.aa.pos)
  top.brd4.list[[con.i]] <- left_join(top.brd4.list[[con.i]], target.aa.pos, by = "Sequence")
  
  # write BRD4 aa table 
  write.table(
    top.brd4.list[[con.i]],
    file.path(edger.dir, paste0(colnames(contr.mat)[con.i], "_BRD4_top_Zscore_aaPos.txt")),
    sep = "\t",
    quote = FALSE,
    col.names = TRUE,
    row.names = FALSE
  )
  
  # make and save protiler input
  protiler.input <- top.brd4.list[[con.i]] %>% 
    dplyr::select(Sequence, Gene, aa.pos, Z.score, neg.Z.score) %>% 
    column_to_rownames("Sequence")
  colnames(protiler.input) <- c("Symbol", "AA", "Z-score", "Z-score_neg")
  write.csv(
    protiler.input,
    file.path(edger.dir, paste0(colnames(contr.mat)[con.i], "_BRD4_protilerInput.txt")),
    quote = FALSE
  )
  
  # get and save top 50 hits
  top.hits <- head(top.brd4.list[[con.i]] %>% arrange(desc(logFC)), 50)
  write.table(
    top.hits,
    file.path(edger.dir, paste0(colnames(contr.mat)[con.i], "_BRD4_top50hits.txt")),
    sep = "\t",
    quote = FALSE,
    col.names = TRUE,
    row.names = FALSE
  )
}
```




## Carry out camera and roast gene-set analysis
```{r}
# get list of sgRNAs per gene
guide.ann$group <- ifelse(
  guide.ann$Gene == "BRD4", 
  paste(guide.ann$Gene, guide.ann$Rank, sep = "_"), 
  guide.ann$Gene
  )
#genesymbols = guide.ann$group
genesymbollist = list()
unq = unique(guide.ann$group)
unq = unq[!is.na(unq)]
for (i in unq) {
  sel = guide.ann$group == i
  if (sum(sel) >= 1){
    genesymbollist[[i]] = which(sel)
  }
}

# run camera and roast for each contrast
for(con.i in 1:length(colnames(contr.mat))){
  # Run camera for all genes
  camera.res = camera(xglm, index = genesymbollist, design, contrast = contr.mat[, con.i], )
  # write out camera results
  write.table(
    camera.res,
    file.path(edger.dir, paste0(colnames(contr.mat)[con.i], "_BRD4_camera_results.txt")),
    sep = "\t",
    quote = FALSE,
    col.names = NA,
    row.names = TRUE
  )
  # Run roast for all genes & write out
  roast.res <- roast(xglm, index = genesymbollist, design, contrast = contr.mat[, con.i])
  roast.res[roast.res$FDR < 0.05, ] %>% filter(Direction == "Up")
  roast.res %>% filter(Direction == "Up" & FDR < 0.1) %>% arrange(desc(PropUp))
  write.table(
    camera.res,
    file.path(edger.dir, paste0(colnames(contr.mat)[con.i], "_BRD4_roast_results.txt")),
    sep = "\t",
    quote = FALSE,
    col.names = NA,
    row.names = TRUE
  )
}
```

## plot Z score, logCPM and FDR across protein BEs body
```{r}
for(con.i in 1:length(colnames(contr.mat))){
  # z-score
  pdf(file.path(plots.dir, paste0("BRD4_pos_Zscore_", colnames(contr.mat)[con.i], ".pdf")))
  top.brd4.list[[con.i]] %>% dplyr::filter(FDR < 0.5) %>% 
    ggplot(aes(x = aa.pos, y = Z.score, color = FDR)) + 
    geom_point() + 
    geom_hline(yintercept = 5, color = "blue") + theme_linedraw() +
    scale_color_continuous(low = "grey20", high = "grey90") +
    ggtitle(colnames(contr.mat)[con.i]) +
    theme_light()
  dev.off()
  # logCPM
  pdf(file.path(plots.dir, paste0("BRD4_pos_logCPM_", colnames(contr.mat)[con.i], ".pdf")))
  top.brd4.list[[con.i]] %>% dplyr::filter(FDR < 0.5) %>% 
    ggplot(aes(x = aa.pos, y = logCPM, color = FDR)) + 
    geom_point() + 
    scale_color_continuous(low = "grey20", high = "grey90") +
    ggtitle(colnames(contr.mat)[con.i]) +
    theme_light()
  dev.off()
  # FDR
  pdf(file.path(plots.dir, paste0("BRD4_pos_FDR_", colnames(contr.mat)[con.i], ".pdf")))
  top.brd4.list[[con.i]] %>% dplyr::filter(FDR < 0.5) %>% 
    ggplot(aes(x = aa.pos, y = FDR, color = FDR)) + 
    geom_point() + 
    scale_color_continuous(low = "grey20", high = "grey90") +
    ggtitle(colnames(contr.mat)[con.i]) +
    ylim(0, 0.1) + 
    theme_light()
  dev.off()
}
```

## draw BRD4 structure using drawProteins package
```{r}
# see tutorial at https://bioconductor.org/packages/devel/bioc/vignettes/drawProteins/inst/doc/drawProteins_BiocStyle.html
# UniProt accession numbers of BRD4 = O60885
rel_json <- drawProteins::get_features("O60885")
rel_data <- drawProteins::feature_to_dataframe(rel_json)
head(rel_data)
```

## plot domains alongside screen data
```{r}
# subset rel data
rel_data %>% filter(type == "CHAIN")
unique(rel_data$type)
prot.data <- rel_data %>% filter(type == "DOMAIN" | type == "REGION" & description != "Disordered")

# region & domain plot
prot.plot <- draw_canvas(rel_data)
prot.plot <- draw_chains(prot.plot, rel_data, label_chains = F) + 
  coord_cartesian(xlim = c(0, 1351))
prot.plot <- draw_regions(prot.plot, prot.data, show.legend = TRUE) + 
  theme(legend.position = "bottom")
prot.plot <- draw_domains(prot.plot, prot.data, label_size = 2) + 
  theme_classic() + 
  theme(legend.position = "bottom", 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.line.y = element_blank()) 

for(con.i in 1:length(colnames(contr.mat))){
  # screen plot
  screen.plot <- top.brd4.list[[con.i]] %>% dplyr::filter(FDR < 0.5) %>% 
      ggplot(aes(x = aa.pos, y = Z.score, color = FDR)) + 
      geom_point() + 
      geom_hline(yintercept = 4, color = "red3") + theme_linedraw() +
      scale_color_continuous(low = "darkblue", high = "grey90") +
      ggtitle(colnames(contr.mat)[con.i]) +
      theme_light() + 
      theme(legend.position = "right")
  screen.plot <- screen.plot + theme(axis.title.x = element_blank())
  
  plt <- list(screen.plot = screen.plot, prot.plot = prot.plot)
  ggpubr::ggarrange(plt$screen.plot, plt$prot.plot, heights = c(12, 4),
            ncol = 1, nrow = 2, align = "v", legend = "bottom", common.legend = TRUE)
  
  ggsave(file.path(plots.dir, paste0("BRD4_prot_Zscore_", colnames(contr.mat)[con.i], ".pdf")), width = 8, height = 5)
}
```



# EDGER DATA
#```{r}
#
#HIGH.camera <- camera.res
#HIGH.roast <- roast.res.high
#
#HIGH.camera <- merge(HIGH.camera, HIGH.roast, by=0, all.x=TRUE)
#
## convert FDR to -log10(FDR)
#HIGH.camera$LogFDR = -log10(HIGH.camera$FDR)
#
#dat <- HIGH.camera
#dat$RandomIndex <- sample(1:nrow(dat), nrow(dat))
#alphabetical <- HIGH.camera[order(rownames(HIGH.camera)),]
#dat$alphabetical <- seq(1,nrow(dat))
#dat <- dat[order(-dat$LogFDR), ]
#pos <- dat[dat$Direction.x == "Up", ]
#pos <- pos %>% mutate(color = if_else(pos$LogFDR>2, true = "Sig", false = "NotSig"))
#neg <- dat[dat$Direction == "Down", ]
#dat <- dat %>% mutate(color = if_else(dat$LogFDR>2,true = "Sig", false = "NotSig"))
##dat$color[which(dat$PropDn < 0.6)] <- "NotSig"
#dat$label <- as.character(rownames(dat))
#dat$label <- ifelse(dat$LogFDR>2, dat$label, "")
##dat$label[which(dat$PropDn < 0.6)] <- ""
#ixlabel <- which(dat$LogFDR>2 & dat$PropDn > 0.6)
#dat$label[ixlabel] <- dat$Row.names[ixlabel] 
#
## Dotplot
#ggplot(dat, aes(RandomIndex, LogFDR, label = label)) + 
#  geom_point(alpha = 0.4) +
#  scale_y_continuous(trans = "log1p") +
#  scale_colour_manual(values = c("grey60", "purple2")) +
#  scale_size_continuous(breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1)) +
#  theme_bw() +
#  #ylim(c(0,3)) +
#  geom_hline(yintercept = 2, color = "black", linetype = 3) +
#  #facet_wrap(~ Timepoint) + 
#  ggrepel::geom_text_repel(size = 4, max.overlaps = 1000)
#  zggsave(filename = "./P300_EdgeR_dotplots.pdf", width = 5.5, height = 3.5)
#
#ggplot(dat, aes(RandomIndex, LogFDR, label = label, size = LogFDR)) + 
#  geom_point(alpha = 0.4) +
#  scale_y_continuous(trans = "log1p") +
#  scale_colour_manual(values = c("grey60", "purple2")) +
#  scale_size_continuous(breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1)) +
#  theme_bw() +
#  ylim(c(0,3)) +
#  geom_hline(yintercept = 1.301, color = "black", linetype = 3) +
#  #facet_wrap(~ Timepoint) + 
#  ggrepel::geom_text_repel(size = 2, max.overlaps = 1000)
#ggsave(filename = "./P300_EdgeR_ylim3_dotplots.pdf", width = 5.5, height = 3.5)
#
#
## 24 vs 48hr scatter
#dat %>% select(X, LogFDR, Timepoint) %>% pivot_wider(values_from = LogFDR, names_from = X, id_cols = Timepoint) %>% column_to_rownames("Timepoint") %>% t() %>% as.data.frame() %>% rownames_to_column() %>% as_tibble() %>% 
#  mutate(labs = ifelse(P300.24h > 1.3 | P300.48h > 1.3, rowname, "")) %>%
#  mutate(sig = ifelse(P300.24h > 1.3 & P300.48h > 1.3, "sig.both", ifelse(P300.24h > 1.3 & P300.48h < 1.3, "sig.24", 
#                             ifelse(P300.24h < 1.3 & P300.48h > 1.3, "sig.48", "not.sig")))) %>%
#  ggplot(aes(x = P300.24h, y = P300.48h, label = labs, color = sig)) + geom_point() +
#  scale_x_continuous(trans = "log1p") +
#  scale_y_continuous(trans = "log1p") +
#  geom_hline(yintercept = 1.3, linetype = 3) +
#  geom_vline(xintercept = 1.3, linetype = 3) +
#  ggrepel::geom_text_repel(max.overlaps = ) + theme_classic() +
#  scale_color_manual(values = c("black", "red", "blue", "green"))
#ggsave(filename = "./P300_EdgeR_dotplot_24_v_48.pdf", width = 4.5, height = 4)
#```

# save session info
```{r}
sesh.info <- capture.output(sessionInfo())
writeLines(
  sesh.info,
  file.path(project.dir, "session-info", paste0(Sys.Date(), "_diff_analysis_240808.txt"))
)
```