---
title: "baseEditor_240529"
author: "Andrea"
date: "2024-06-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

fastq dir: /pipeline/Runs/NextSeq/240529_VH01624_107_AAFLMTJM5/ProjectFolders/Project_Oliver-Sinclair/Sample*
project dir: /dawson_genomics/Projects/CCS1477/baseEditor_240529

# make bowtie2 index from fasta of Ollie's guide lib list (done interactively)
```{bash}
module load bowtie2/2.3.4.1

cd reference_files
bowtie2-build -f Guides_BRD4_VHL.fa Guides_BRD4_VHL
```


# align base editor library to guide library fasta from list sent by Ollie
# aligning with 0 mismatches to start 
```{bash}
mkdir fastqc
mkdir scripts
mkdir logs
mkdir trimmed_fastqs
mkdir alignment
mkdir session-info
mkdir mapped_guides
mkdir guide_counts

# R2 is all G's indicating inability to sequence, so only align R1
sbatch scripts/MapCRISPR.sbatch /pipeline/Runs/NextSeq/240529_VH01624_107_AAFLMTJM5/ProjectFolders/Project_Oliver-Sinclair/Sample_BRD4-lib/BRD4-lib_S1_R1_001.fastq.gz

module load multiqc/1.8
multiqc .
```

# get stats for alignment
```{bash}
module load samtools/1.17

samtools flagstat alignment/BRD4-lib_S1.headed.sam 
```
3536919 + 0 in total (QC-passed reads + QC-failed reads)
3536919 + 0 primary
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
3332491 + 0 mapped (94.22% : N/A)
3332491 + 0 primary mapped (94.22% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)

# load libraries and set paths
```{r}
library(data.table)
library(ggplot2)

project.dir <- "/dawson_genomics/Projects/CCS1477/baseEditor_240529"
plots.dir <- file.path(project.dir, "plots")
```

# make a density plot of each guide type
```{r}
guide.counts <- fread(file.path(project.dir, "guide_counts/BRD4-lib_S1.guideCounts.txt"))

# remove unaligned counts
guide.counts <- guide.counts[-1, ]
guide.counts$category <- gsub("_", "", substr(guide.counts$V2, 1, 4))
colnames(guide.counts)[1] <- "counts" 

pdf(file.path(plots.dir, "density_guideCounts.pdf"))
ggplot(guide.counts, aes(counts, colour = category, fill = category)) +
  geom_density(alpha = 0.3) + 
  theme_classic()
dev.off()
```


# save session info
```{r}
sesh.info <- capture.output(sessionInfo())
writeLines(
  sesh.info,
  file.path(project.dir, "session-info", paste0(Sys.Date(), "_baseEditor_240529.txt"))
)
```
